File: samples/CameraAccess/CameraAccess/Settings/SettingsManager.swift
import Foundation

enum AgentBackend: String, CaseIterable {
  case e2b = "E2B"
  case openClaw = "OpenClaw"
  case studioVaultMCP = "StudioVaultMCP"   // Phase 2 (StudioVault-Glasses fork): HostBroker-routed vault ops
}

/// Voice provider selector for the StudioVault-Glasses fork.
/// Added to support Azure Realtime alongside the original Gemini Live provider.
enum VoiceProvider: String, CaseIterable {
  case geminiLive = "Gemini Live"
  case azureRealtime = "Azure Realtime"
}

final class SettingsManager {
  static let shared = SettingsManager()

  private let defaults = UserDefaults.standard

  private enum Key: String {
    case geminiAPIKey
    case geminiSystemPrompt
    case azureOpenAIAPIKey
    case azureRealtimeBase
    case azureRealtimeDeployment
    case azureRealtimeSystemPrompt
    case hostBrokerURL
    case hostBrokerToken
    case voiceProvider
    case agentBackend
    case agentBaseURL
    case agentToken
    case openClawHost
    case openClawPort
    case openClawHookToken
    case openClawGatewayToken
    case webrtcSignalingURL
    case speakerOutputEnabled
    case videoStreamingEnabled
    case proactiveNotificationsEnabled
    case userId
    case agentSessionKey
    case agentSessionCreatedAt
    case fontTheme
  }

  private init() {
    migrateSystemPromptIfNeeded()
  }

  /// One-time migration: replace the old "you have NO ability" system prompt
  /// with the new positive-framing default so Gemini uses tools confidently.
  private func migrateSystemPromptIfNeeded() {
    guard let stored = defaults.string(forKey: Key.geminiSystemPrompt.rawValue),
          stored.contains("You have NO memory, NO storage, and NO ability") else { return }
    defaults.removeObject(forKey: Key.geminiSystemPrompt.rawValue)
    NSLog("[Settings] Migrated system prompt to new default")
  }

  // MARK: - Gemini

  var geminiAPIKey: String {
    get { defaults.string(forKey: Key.geminiAPIKey.rawValue) ?? Secrets.geminiAPIKey }
    set { defaults.set(newValue, forKey: Key.geminiAPIKey.rawValue) }
  }

  var geminiSystemPrompt: String {
    get { defaults.string(forKey: Key.geminiSystemPrompt.rawValue) ?? GeminiConfig.defaultSystemInstruction }
    set { defaults.set(newValue, forKey: Key.geminiSystemPrompt.rawValue) }
  }

  // MARK: - Azure OpenAI Realtime (StudioVault-Glasses fork)

  var azureOpenAIAPIKey: String {
    get { defaults.string(forKey: Key.azureOpenAIAPIKey.rawValue) ?? Secrets.azureOpenAIAPIKey }
    set { defaults.set(newValue, forKey: Key.azureOpenAIAPIKey.rawValue) }
  }

  var azureRealtimeBase: String {
    get { defaults.string(forKey: Key.azureRealtimeBase.rawValue) ?? Secrets.azureRealtimeBase }
    set { defaults.set(newValue, forKey: Key.azureRealtimeBase.rawValue) }
  }

  var azureRealtimeDeployment: String {
    get { defaults.string(forKey: Key.azureRealtimeDeployment.rawValue) ?? Secrets.azureRealtimeDeployment }
    set { defaults.set(newValue, forKey: Key.azureRealtimeDeployment.rawValue) }
  }

  var azureRealtimeSystemPrompt: String {
    get { defaults.string(forKey: Key.azureRealtimeSystemPrompt.rawValue) ?? AzureRealtimeConfig.defaultSystemInstruction }
    set { defaults.set(newValue, forKey: Key.azureRealtimeSystemPrompt.rawValue) }
  }

  // MARK: - Voice provider selector (StudioVault-Glasses fork)

  var voiceProvider: VoiceProvider {
    get {
      guard let raw = defaults.string(forKey: Key.voiceProvider.rawValue),
            let provider = VoiceProvider(rawValue: raw) else { return .geminiLive }
      return provider
    }
    set { defaults.set(newValue.rawValue, forKey: Key.voiceProvider.rawValue) }
  }

  // MARK: - HostBroker (StudioVault-Glasses fork, Phase 2)

  var hostBrokerURL: String {
    get { defaults.string(forKey: Key.hostBrokerURL.rawValue) ?? Secrets.hostBrokerURL }
    set { defaults.set(newValue, forKey: Key.hostBrokerURL.rawValue) }
  }

  var hostBrokerToken: String {
    get { defaults.string(forKey: Key.hostBrokerToken.rawValue) ?? Secrets.hostBrokerToken }
    set { defaults.set(newValue, forKey: Key.hostBrokerToken.rawValue) }
  }

  // MARK: - Agent

  var agentBackend: AgentBackend {
    get {
      guard let raw = defaults.string(forKey: Key.agentBackend.rawValue),
            let backend = AgentBackend(rawValue: raw) else { return .e2b }
      return backend
    }
    set { defaults.set(newValue.rawValue, forKey: Key.agentBackend.rawValue) }
  }

  // E2B settings
  var agentBaseURL: String {
    get { defaults.string(forKey: Key.agentBaseURL.rawValue) ?? Secrets.agentBaseURL }
    set { defaults.set(newValue, forKey: Key.agentBaseURL.rawValue) }
  }

  var agentToken: String {
    get { defaults.string(forKey: Key.agentToken.rawValue) ?? Secrets.agentToken }
    set { defaults.set(newValue, forKey: Key.agentToken.rawValue) }
  }

  // OpenClaw settings
  var openClawHost: String {
    get { defaults.string(forKey: Key.openClawHost.rawValue) ?? Secrets.openClawHost }
    set { defaults.set(newValue, forKey: Key.openClawHost.rawValue) }
  }

  var openClawPort: Int {
    get {
      let stored = defaults.integer(forKey: Key.openClawPort.rawValue)
      return stored != 0 ? stored : Secrets.openClawPort
    }
    set { defaults.set(newValue, forKey: Key.openClawPort.rawValue) }
  }

  var openClawHookToken: String {
    get { defaults.string(forKey: Key.openClawHookToken.rawValue) ?? Secrets.openClawHookToken }
    set { defaults.set(newValue, forKey: Key.openClawHookToken.rawValue) }
  }

  var openClawGatewayToken: String {
    get { defaults.string(forKey: Key.openClawGatewayToken.rawValue) ?? Secrets.openClawGatewayToken }
    set { defaults.set(newValue, forKey: Key.openClawGatewayToken.rawValue) }
  }

  // MARK: - WebRTC

  var webrtcSignalingURL: String {
    get { defaults.string(forKey: Key.webrtcSignalingURL.rawValue) ?? Secrets.webrtcSignalingURL }
    set { defaults.set(newValue, forKey: Key.webrtcSignalingURL.rawValue) }
  }

  // MARK: - Audio

  var speakerOutputEnabled: Bool {
    get { defaults.bool(forKey: Key.speakerOutputEnabled.rawValue) }
    set { defaults.set(newValue, forKey: Key.speakerOutputEnabled.rawValue) }
  }

  var videoStreamingEnabled: Bool {
    get {
      // Default to true if never set
      if defaults.object(forKey: Key.videoStreamingEnabled.rawValue) == nil { return true }
      return defaults.bool(forKey: Key.videoStreamingEnabled.rawValue)
    }
    set { defaults.set(newValue, forKey: Key.videoStreamingEnabled.rawValue) }
  }

  var proactiveNotificationsEnabled: Bool {
    get { defaults.bool(forKey: Key.proactiveNotificationsEnabled.rawValue) }
    set { defaults.set(newValue, forKey: Key.proactiveNotificationsEnabled.rawValue) }
  }

  // MARK: - User Identity (permanent, survives reset)

  var userId: String {
    if let id = defaults.string(forKey: Key.userId.rawValue) { return id }
    let id = UUID().uuidString
    defaults.set(id, forKey: Key.userId.rawValue)
    return id
  }

  // MARK: - Font Theme

  var fontTheme: String {
    get { defaults.string(forKey: Key.fontTheme.rawValue) ?? FontTheme.tiempos.rawValue }
    set { defaults.set(newValue, forKey: Key.fontTheme.rawValue) }
  }

  // MARK: - Session Persistence

  var agentSessionKey: String? {
    get { defaults.string(forKey: Key.agentSessionKey.rawValue) }
    set { defaults.set(newValue, forKey: Key.agentSessionKey.rawValue) }
  }

  var agentSessionCreatedAt: TimeInterval {
    get { defaults.double(forKey: Key.agentSessionCreatedAt.rawValue) }
    set { defaults.set(newValue, forKey: Key.agentSessionCreatedAt.rawValue) }
  }

  // MARK: - Reset

  func resetAll() {
    for key in [Key.geminiAPIKey, .geminiSystemPrompt, .agentBackend,
                .agentBaseURL, .agentToken,
                .openClawHost, .openClawPort, .openClawHookToken, .openClawGatewayToken,
                .azureOpenAIAPIKey, .azureRealtimeBase, .azureRealtimeDeployment, .azureRealtimeSystemPrompt,
                .hostBrokerURL, .hostBrokerToken, .voiceProvider,
                .webrtcSignalingURL, .speakerOutputEnabled, .videoStreamingEnabled, .proactiveNotificationsEnabled,
                .agentSessionKey, .agentSessionCreatedAt] {
      defaults.removeObject(forKey: key.rawValue)
    }
  }
}

File: samples/CameraAccessAndroid/app/src/main/java/com/meta/wearable/dat/externalsampleapps/cameraaccess/settings/SettingsManager.kt
package com.meta.wearable.dat.externalsampleapps.cameraaccess.settings

import android.content.Context
import android.content.SharedPreferences
import android.util.Log
import com.meta.wearable.dat.externalsampleapps.cameraaccess.Secrets

enum class AgentBackend(val label: String) {
    E2B("E2B"),
    OPENCLAW("OpenClaw");

    companion object {
        fun fromLabel(label: String): AgentBackend =
            entries.find { it.label == label } ?: E2B
    }
}

object SettingsManager {
    private const val TAG = "SettingsManager"
    private const val PREFS_NAME = "visionclaw_settings"

    private lateinit var prefs: SharedPreferences

    fun init(context: Context) {
        prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
        migrateSystemPromptIfNeeded()
    }

    /** One-time migration: replace old negative-framing system prompt with new positive default */
    private fun migrateSystemPromptIfNeeded() {
        val stored = prefs.getString("geminiSystemPrompt", null) ?: return
        if (stored.contains("You have NO memory, NO storage, and NO ability")) {
            prefs.edit().remove("geminiSystemPrompt").apply()
            Log.d(TAG, "Migrated system prompt to new default")
        }
    }

    // -- Gemini --

    var geminiAPIKey: String
        get() = prefs.getString("geminiAPIKey", null) ?: Secrets.geminiAPIKey
        set(value) = prefs.edit().putString("geminiAPIKey", value).apply()

    var geminiSystemPrompt: String
        get() = prefs.getString("geminiSystemPrompt", null) ?: DEFAULT_SYSTEM_PROMPT
        set(value) = prefs.edit().putString("geminiSystemPrompt", value).apply()

    // -- Agent Backend --

    var agentBackend: AgentBackend
        get() {
            val raw = prefs.getString("agentBackend", null) ?: return AgentBackend.E2B
            return AgentBackend.fromLabel(raw)
        }
        set(value) = prefs.edit().putString("agentBackend", value.label).apply()

    // -- E2B --

    var agentBaseURL: String
        get() = prefs.getString("agentBaseURL", null) ?: Secrets.agentBaseURL
        set(value) = prefs.edit().putString("agentBaseURL", value).apply()

    var agentToken: String
        get() = prefs.getString("agentToken", null) ?: Secrets.agentToken
        set(value) = prefs.edit().putString("agentToken", value).apply()

    // -- OpenClaw --

    var openClawHost: String
        get() = prefs.getString("openClawHost", null) ?: Secrets.openClawHost
        set(value) = prefs.edit().putString("openClawHost", value).apply()

    var openClawPort: Int
        get() {
            val stored = prefs.getInt("openClawPort", 0)
            return if (stored != 0) stored else Secrets.openClawPort
        }
        set(value) = prefs.edit().putInt("openClawPort", value).apply()

    var openClawHookToken: String
        get() = prefs.getString("openClawHookToken", null) ?: Secrets.openClawHookToken
        set(value) = prefs.edit().putString("openClawHookToken", value).apply()

    var openClawGatewayToken: String
        get() = prefs.getString("openClawGatewayToken", null) ?: Secrets.openClawGatewayToken
        set(value) = prefs.edit().putString("openClawGatewayToken", value).apply()

    // -- WebRTC --

    var webrtcSignalingURL: String
        get() = prefs.getString("webrtcSignalingURL", null) ?: Secrets.webrtcSignalingURL
        set(value) = prefs.edit().putString("webrtcSignalingURL", value).apply()

    fun resetAll() {
        prefs.edit().clear().apply()
    }

    const val DEFAULT_SYSTEM_PROMPT = """You are an AI assistant for someone wearing Meta Ray-Ban smart glasses. You can see through their camera and have a voice conversation. Keep responses concise and natural.

You have powerful tools that let you take real actions in the world. Use them confidently whenever the user asks for help.

TOOLS:

1. execute -- Your main tool. It connects to a personal assistant that can do anything:
- Send messages (WhatsApp, Telegram, iMessage, Slack, email, etc.)
- Search the web, look up facts, news, local info
- Check and manage email, calendar, notes, reminders, todos
- Create, edit, or organize documents and files
- Research and analyze topics
- Remember information for later
- Control apps, devices, and services

2. capture_photo -- Capture and save the current camera frame as a photo. Use when the user asks to take a photo, capture what you see, save a picture, or snap a photo. You can optionally include a brief description. This works instantly.

RULES:

- When the user asks you to do something actionable, ALWAYS use execute. Pass a detailed task description with all relevant context (names, content, platforms, quantities, etc.).
- NEVER say you can't do something that execute can handle. If in doubt, try it.
- NEVER pretend to complete an action without actually calling the tool.
- Before calling execute, ALWAYS speak a brief acknowledgment first so the user knows you heard them. For example:
  - "Sure, let me check your email." then call execute.
  - "Got it, searching for that now." then call execute.
  - "On it, sending that message." then call execute.
- The tool may take several seconds, so the verbal acknowledgment is important.
- For messages, confirm recipient and content before sending unless clearly urgent."""
}
